### ISSUE TITLE
Enemy hitboxes and headshots.

### ISSUE THREAD

**Issue by r3538987** [Realm667/WolfenDoom#1319]
Had anyone encountered such issue, where headshots simply doesn't work. You shoot, and bullets to through head, while body shots register properly.
What parameters should I check, maybe this can be tweaked somewhere?
sample 1 - https://webmshare.com/play/8KXrZ
sample 2 - https://webmshare.com/play/GeVv7

Who knows, maybe this is intended, but after Brutal Wolfenstein feels very odd when something is different with headshots.

**Comment by Username-N00b-is-not-available** [comment]
Can confirm (the default Guard height is 64):
![image](https://github.com/Realm667/WolfenDoom/assets/59409713/5a6e14c9-944b-455e-95fd-eda0d7552241)
Looks like the actor kneels when out of player sight and changes his height to 0.75 of original as in this line:
https://github.com/Realm667/WolfenDoom/blob/4cd023bcf0614a834aa24b4cebe84bafca1bb860/scripts/actors/enemies/base.zs#L1690
but this change does not get reverted.
@r3538987 Thanks for noticing!

**Comment by r3538987** [comment]
Any hint how I can repack main ipk3 on my own? Because it doesn't feel that we are getting another release, based on how long recent one was made.

**Comment by Username-N00b-is-not-available** [comment]
There's a dedicated tool in the repository (it's just a batch script + 7zip) for repacking the ipk3, namely **build.bat** (or .sh on Linux).
A fix would be to call the same `A_SetSize` function with just the `default.height` at the beginning of each `See` state for regular soldiers (those that can crouch, i. e. `See.Normal`, `See.Fast`, `See.Faster`, `See.Stand`, and possibly `See.Frightened`), however I'm not sure that wouldn't break any other existing features, and I can't test right now.

**Comment by r3538987** [comment]
Okay, I can wait for proper fix. Because to fix on my own probably won't work, doubt that it requires editing single string. 
@Username-N00b-is-not-available  you also mentioned regular soldier, but on mine second sample there is higher tied person for sure. Or by game's logic he is still regular one too? And other are like dogs/bosses/...

**Comment by Username-N00b-is-not-available** [comment]
Seems to be fixed in my upcoming commit (will do it later today, working on other issues currently):
![image](https://github.com/Realm667/WolfenDoom/assets/59409713/0ac78b65-f45f-41c1-8922-d2494afd8f60)

The officers use the `See.Faster` state, and you are indeed right in the last part of your last message.
Thanks again a lot for catching this subtle bug! The Nazis do not receive additional headshot damage, so it seems like neither the devs nor playtesters ever tried to aim at anything other than the chest/stomach :-)

**Comment by Username-N00b-is-not-available** [comment]
Added the fix in the new commit. For this specific line, it's been in code since 11 Jan 2019 :-)
I remember noticing that bullets sometimes pass through an enemy if I accidentally shoot them in the head, and thinking that was some arcane bug that happens as a result of multiple side effects (floorclip, slopes, current star formation on the sky and so on). It turned out that the explanation was simple, I hope my commit does, in fact, fix this...
This happened mostly because in BoA it makes little sense to prioritize headshots (has no impact on damage + you are more likely to miss the enemy completely by overshooting). You said you're accustomed to BW, and you thus have a reflex 'see Nazi, aim at head' :-) , which none of us had. Still it was an important bug.

**Comment by r3538987** [comment]
This is bit of offtop.
Mine game has 1.02 GB, boa.ipk3 file.
But when copied repository, build Default or No Compression, I get PK3 file (without i), and it's ~700MB only. 

And when tried to replace ipk3 with pk3, but renaming it. 
GZDoom on load throws this:

```
Execution could not continue.
Script error, "boa.ipk3:mapinfo.txt" line 59:
Expected '=' but got '{' instead.
```

**Comment by Username-N00b-is-not-available** [comment]
Yes, this is because of recent additions (more recent than 2021 at least). The contents of the repository are tested on the latest GZDoom development build.
I had also noticed the filesize difference, not sure what caused it.

**Comment by r3538987** [comment]
Seems it will be long journey. 
GZDoom 4.10.0
`Script error, "boa.ipk3:scripts/actors/effects/clouds.zs" line 103:
Too many arguments in call to Trace
Script error, "boa.ipk3:scripts/actors/effects/snow.zs" line 88:
Too many arguments in call to Trace
`
Wont be surprised if save files also will fail :)

**Comment by Username-N00b-is-not-available** [comment]
You are right, the saves would be incompatible, since there were a lot of map changes and fixes during these 2 years.
If you planned to continue the playthrough, you'd be better off with old .ipk3 and old GZDoom that were in the archive. If, on the other hand, you have finished only a few Chapter 1 levels, then you probably can quickly replay these in a new version (or just change the map via console).

You launched one the release versions, and we are using automated devbuilds that can be downloaded [here](https://devbuilds.drdteam.org/gzdoom/) (note that these are even more prone to bugs than BoA :-) ) I'm currently using `pre-195`, it seems to work fine.

Heh, nice to know that the GZDoom developers has already repaired the automated build system on DRDTeam (it was down for a few months, had to download these builds from GitHub Actions).

**Comment by r3538987** [comment]
Can confirm, this setup works indeed. 
Quite a journey I got. 

Nice that game also keeps tracking progress despite I jumped straight into C1M3. 
Hope nothing major will break without few first maps. (Think only doctor will fail to get his research documents). 

And what does Debug button on main location map? Few times pressed, didn't noticed anything, save loaded, because was stared maybe can break with it.

Also as some feedback - I would say that game more frequently gets noticeable slowdowns. Especially when autosaving. 
Older 2021 build felt like more smooth in this regard.

**Comment by Username-N00b-is-not-available** [comment]
Yes, the lack of secret files is the main problem an `idclev`ing player will face in the progression, however if you are not on Mein Leben (i. e. cheats are on), you can `summon AktenEisenmann`. Plus you're losing coins that were on the levels, but actually not too much of them.
As for the Debug button, it opens a hidden passage in the wall to the right of the achievement board that allows to change levels. It might be obsolete, `map` also works fine.
The performance is indeed one of the most serious problems of current BoA, and high-end hardware for some reason ends up performing worse than older or low-end computers. For now I can only advise to use predefined settings in the menus of 'Medium' or 'Low' quality.

**Comment by r3538987** [comment]
> I had also noticed the filesize difference, not sure what caused it.
Sorry for bumping this, but just out of curiosity wanted to check if really something missing between 700MB and 1GB files.
And after some checks it looks like nothing major is missing, probably that result has to do with some fancy compression which shrinks file this good.

**Comment by r3538987** [comment]
Can we reopen this issue? Because BOSSES probably suffer from similar problem.
Here how it looks on my end - https://webmshare.com/play/q1ZWQ
Player shoots enemy's head, and HP values doesn't decrease. 
With this particular enemy also you can observe that aiming for head hides health bar, meaning by games logic player's targets not into enemy. 

With other bosses I felt similar issue, but due to furious running around can't surely confirm it.

**Comment by x3003** [comment]
>Because BOSSES probably suffer from similar problem.

I've just tested it w/ the boss in c3m5_c and it seems to work. So maybe a specific boss on a specific level that doen't work ?

**Comment by r3538987** [comment]
> I've just tested it w/ the boss in c3m5_c and it seems to work. So maybe a specific boss on a specific level that doen't work ?

Think so, because one showed definitely is broken, but others in majority of time looks fine.

**Comment by Username-N00b-is-not-available** [comment]
The `AirshipGeneral` that is seen on C3M5_C has height 80, as by default have the bosses that are inherited from `NaziBoss`:
https://github.com/Realm667/WolfenDoom/blob/ffc1834f7f8b6ba18bd3286f8aee313d0f4989a5/scripts/actors/enemies/base.zs#L3765
The super soldiers, however, had height 64, as in here:
https://github.com/Realm667/WolfenDoom/blob/ffc1834f7f8b6ba18bd3286f8aee313d0f4989a5/scripts/decorate/monsters/bosses.zs#L3371
Looks like this was just a leftover from the time when the hi-res boss sprites were added (around 24--25 Apr 2017). I committed the changes. @r3538987, this time you'll only have to rebuild the pk3 from the repository, the savegames will not be broken (as is the case with ZScript changes that don't do anything major, like removing classes).

**Comment by r3538987** [comment]
Hope the last one. 
There are these turret guys, who switch their states from being sideways for player to watching you in the eyes. 
I cannot come with suggestion how this can be improved. Like when you peak these guys, they are sideways, you shoot into them, and after tiny delay they switch position and your bullet now flies into nothing.

Now and then also felt problems shooting these dudes, but here was another sample where such thing became more obvious. 
https://webmshare.com/play/KbKyv

**Comment by r3538987** [comment]
https://webmshare.com/play/gwm75
Seems like barrels also want some piece of action.

**Comment by Tormentor667** [comment]
Is this already live or is it part of a fork currently? Sorry for derailing, I totally missed this.

**Comment by r3538987** [comment]
Slight bump.
Just wanted to close this ticket, becauase was scrolling Moddb and noticed that BOA file on site was updated on Feb 8th, 2026.
So gave another shot on first level and noticed that bullets still go through enemy heads. 

**Could it be that while file was "Updated", its still based on old version?** I see that ipk3 is from 2021. 

---
Will take latest build and will see how it goes paired with UZDoom. 
UZDoom v4.14.3 & commit https://github.com/Realm667/WolfenDoom/commit/2bc00b027bab020d94cce509e366d9229d1921fc plays nicely. But turret guys still cannot receive headshots. 

https://github.com/user-attachments/assets/4a617ec9-1385-4091-bc4f-e76919d94f6b

**Comment by hawkwind3** [comment]
@r3538987 - Fixed. Can now do head shots. Thing Types 12790 and 12787 had too low height. Was 56. Changed to 64. Also increased height of controllable turret to 40. Was 32.
